File: src/renderer/src/services/ipc/sftp.ts
import { backendService } from './backend'
import { FileInfo, TransferProgress, IPCMessage } from '../../types'

export const sftpService = {
  list(sessionId: string, path: string): Promise<FileInfo[]> {
    return new Promise((resolve, reject) => {
      const handler = (message: IPCMessage) => {
        if (message.session_id === sessionId && message.type === 'sftp:list') {
          backendService.off('sftp:list')
          resolve(message.data as FileInfo[])
        } else if (message.type === 'error') {
          backendService.off('error')
          reject(new Error(message.data.error))
        }
      }

      backendService.on('sftp:list', handler)
      backendService.on('error', handler)

      backendService.send({
        type: 'sftp:list',
        session_id: sessionId,
        data: { path }
      })
    })
  },

  upload(
    sessionId: string,
    localPath: string,
    remotePath: string,
    onProgress?: (progress: TransferProgress) => void
  ): Promise<void> {
    return new Promise((resolve, reject) => {
      if (onProgress) {
        const progressHandler = (message: IPCMessage) => {
          if (message.session_id === sessionId && message.type === 'sftp:progress') {
            onProgress(message.data as TransferProgress)
          }
        }
        backendService.on('sftp:progress', progressHandler)
      }

      const handler = (message: IPCMessage) => {
        if (message.session_id === sessionId && message.type === 'sftp:upload') {
          backendService.off('sftp:upload')
          backendService.off('sftp:progress')
          resolve()
        } else if (message.type === 'error') {
          backendService.off('error')
          backendService.off('sftp:progress')
          reject(new Error(message.data.error))
        }
      }

      backendService.on('sftp:upload', handler)
      backendService.on('error', handler)

      backendService.send({
        type: 'sftp:upload',
        session_id: sessionId,
        data: { local_path: localPath, remote_path: remotePath }
      })
    })
  },

  download(
    sessionId: string,
    remotePath: string,
    localPath: string,
    onProgress?: (progress: TransferProgress) => void
  ): Promise<void> {
    return new Promise((resolve, reject) => {
      if (onProgress) {
        const progressHandler = (message: IPCMessage) => {
          if (message.session_id === sessionId && message.type === 'sftp:progress') {
            onProgress(message.data as TransferProgress)
          }
        }
        backendService.on('sftp:progress', progressHandler)
      }

      const handler = (message: IPCMessage) => {
        if (message.session_id === sessionId && message.type === 'sftp:download') {
          backendService.off('sftp:download')
          backendService.off('sftp:progress')
          resolve()
        } else if (message.type === 'error') {
          backendService.off('error')
          backendService.off('sftp:progress')
          reject(new Error(message.data.error))
        }
      }

      backendService.on('sftp:download', handler)
      backendService.on('error', handler)

      backendService.send({
        type: 'sftp:download',
        session_id: sessionId,
        data: { remote_path: remotePath, local_path: localPath }
      })
    })
  },

  delete(sessionId: string, path: string): Promise<void> {
    return new Promise((resolve, reject) => {
      const handler = (message: IPCMessage) => {
        if (message.session_id === sessionId && message.type === 'sftp:delete') {
          backendService.off('sftp:delete')
          resolve()
        } else if (message.type === 'error') {
          backendService.off('error')
          reject(new Error(message.data.error))
        }
      }

      backendService.on('sftp:delete', handler)
      backendService.on('error', handler)

      backendService.send({
        type: 'sftp:delete',
        session_id: sessionId,
        data: { path }
      })
    })
  },

  mkdir(sessionId: string, path: string): Promise<void> {
    return new Promise((resolve, reject) => {
      const handler = (message: IPCMessage) => {
        if (message.session_id === sessionId && message.type === 'sftp:mkdir') {
          backendService.off('sftp:mkdir')
          resolve()
        } else if (message.type === 'error') {
          backendService.off('error')
          reject(new Error(message.data.error))
        }
      }

      backendService.on('sftp:mkdir', handler)
      backendService.on('error', handler)

      backendService.send({
        type: 'sftp:mkdir',
        session_id: sessionId,
        data: { path }
      })
    })
  },

  rename(sessionId: string, oldPath: string, newPath: string): Promise<void> {
    return new Promise((resolve, reject) => {
      const handler = (message: IPCMessage) => {
        if (message.session_id === sessionId && message.type === 'sftp:rename') {
          backendService.off('sftp:rename')
          resolve()
        } else if (message.type === 'error') {
          backendService.off('error')
          reject(new Error(message.data.error))
        }
      }

      backendService.on('sftp:rename', handler)
      backendService.on('error', handler)

      backendService.send({
        type: 'sftp:rename',
        session_id: sessionId,
        data: { old_path: oldPath, new_path: newPath }
      })
    })
  },

  cancel(sessionId: string, transferId: string): Promise<boolean> {
    return new Promise((resolve) => {
      const handler = (message: IPCMessage) => {
        if (message.type === 'sftp:cancel') {
          backendService.off('sftp:cancel')
          resolve(message.data?.cancelled ?? false)
        }
      }

      backendService.on('sftp:cancel', handler)

      backendService.send({
        type: 'sftp:cancel',
        session_id: sessionId,
        data: { transfer_id: transferId }
      })
    })
  },

  readFile(sessionId: string, path: string, binary?: boolean): Promise<string> {
    return new Promise((resolve, reject) => {
      console.log('[SFTP] readFile request:', { sessionId, path, binary })
      
      const handler = (message: IPCMessage) => {
        console.log('[SFTP] readFile received message:', message.type, message.session_id)
        if (message.session_id === sessionId && message.type === 'sftp:readfile') {
          console.log('[SFTP] readFile success')
          backendService.off('sftp:readfile')
          backendService.off('error')
          resolve(message.data.content as string)
        } else if (message.session_id === sessionId && message.type === 'error') {
          console.log('[SFTP] readFile error:', message.data)
          backendService.off('sftp:readfile')
          backendService.off('error')
          reject(new Error(message.data.error))
        }
      }

      backendService.on('sftp:readfile', handler)
      backendService.on('error', handler)

      backendService.send({
        type: 'sftp:readfile',
        session_id: sessionId,
        data: { path, binary: binary ?? false }
      })
    })
  },

  writeFile(sessionId: string, path: string, content: string): Promise<void> {
    return new Promise((resolve, reject) => {
      const handler = (message: IPCMessage) => {
        if (message.session_id === sessionId && message.type === 'sftp:writefile') {
          backendService.off('sftp:writefile')
          resolve()
        } else if (message.type === 'error') {
          backendService.off('error')
          reject(new Error(message.data.error))
        }
      }

      backendService.on('sftp:writefile', handler)
      backendService.on('error', handler)

      backendService.send({
        type: 'sftp:writefile',
        session_id: sessionId,
        data: { path, content }
      })
    })
  }
}

File: backend/internal/ipc/handlers/sftp.go
package handlers

import (
	"encoding/json"
	"fmt"
	"freessh-backend/internal/models"
	"freessh-backend/internal/session"
)

type SFTPHandler struct {
	manager *session.Manager
}

func NewSFTPHandler(manager *session.Manager) *SFTPHandler {
	return &SFTPHandler{
		manager: manager,
	}
}

func (h *SFTPHandler) CanHandle(msgType models.MessageType) bool {
	switch msgType {
	case models.MsgSFTPList, models.MsgSFTPUpload, models.MsgSFTPDownload,
		models.MsgSFTPDelete, models.MsgSFTPMkdir, models.MsgSFTPRename, 
		models.MsgSFTPCancel, models.MsgSFTPReadFile, models.MsgSFTPWriteFile,
		models.MsgSFTPChmod:
		return true
	}
	return false
}

func (h *SFTPHandler) Handle(msg *models.IPCMessage, writer ResponseWriter) error {
	switch msg.Type {
	case models.MsgSFTPList:
		return h.handleList(msg, writer)
	case models.MsgSFTPUpload:
		return h.handleUpload(msg, writer)
	case models.MsgSFTPDownload:
		return h.handleDownload(msg, writer)
	case models.MsgSFTPDelete:
		return h.handleDelete(msg, writer)
	case models.MsgSFTPMkdir:
		return h.handleMkdir(msg, writer)
	case models.MsgSFTPRename:
		return h.handleRename(msg, writer)
	case models.MsgSFTPCancel:
		return h.handleCancel(msg, writer)
	case models.MsgSFTPReadFile:
		return h.handleReadFile(msg, writer)
	case models.MsgSFTPWriteFile:
		return h.handleWriteFile(msg, writer)
	case models.MsgSFTPChmod:
		return h.handleChmod(msg, writer)
	default:
		return fmt.Errorf("unsupported message type: %s", msg.Type)
	}
}

func (h *SFTPHandler) handleList(msg *models.IPCMessage, writer ResponseWriter) error {
	jsonData, err := json.Marshal(msg.Data)
	if err != nil {
		return fmt.Errorf("invalid data: %w", err)
	}

	var req models.ListRequest
	if err := json.Unmarshal(jsonData, &req); err != nil {
		return fmt.Errorf("failed to parse list request: %w", err)
	}

	files, err := h.manager.ListFiles(msg.SessionID, req.Path)
	if err != nil {
		return err
	}

	return writer.WriteMessage(&models.IPCMessage{
		Type:      models.MsgSFTPList,
		SessionID: msg.SessionID,
		Data:      files,
	})
}

func (h *SFTPHandler) handleUpload(msg *models.IPCMessage, writer ResponseWriter) error {
	jsonData, err := json.Marshal(msg.Data)
	if err != nil {
		return fmt.Errorf("invalid data: %w", err)
	}

	var req models.UploadRequest
	if err := json.Unmarshal(jsonData, &req); err != nil {
		return fmt.Errorf("failed to parse upload request: %w", err)
	}

	progressChan := make(chan models.TransferProgress, 10)
	
	go func() {
		for progress := range progressChan {
			writer.WriteMessage(&models.IPCMessage{
				Type:      models.MsgSFTPProgress,
				SessionID: msg.SessionID,
				Data:      progress,
			})
		}
	}()

	err = h.manager.UploadFile(msg.SessionID, req.LocalPath, req.RemotePath, progressChan)
	close(progressChan)

	if err != nil {
		return err
	}

	return writer.WriteMessage(&models.IPCMessage{
		Type:      models.MsgSFTPUpload,
		SessionID: msg.SessionID,
		Data:      map[string]string{"status": "completed"},
	})
}

func (h *SFTPHandler) handleDownload(msg *models.IPCMessage, writer ResponseWriter) error {
	jsonData, err := json.Marshal(msg.Data)
	if err != nil {
		return fmt.Errorf("invalid data: %w", err)
	}

	var req models.DownloadRequest
	if err := json.Unmarshal(jsonData, &req); err != nil {
		return fmt.Errorf("failed to parse download request: %w", err)
	}

	progressChan := make(chan models.TransferProgress, 10)
	
	go func() {
		for progress := range progressChan {
			writer.WriteMessage(&models.IPCMessage{
				Type:      models.MsgSFTPProgress,
				SessionID: msg.SessionID,
				Data:      progress,
			})
		}
	}()

	err = h.manager.DownloadFile(msg.SessionID, req.RemotePath, req.LocalPath, progressChan)
	close(progressChan)

	if err != nil {
		return err
	}

	return writer.WriteMessage(&models.IPCMessage{
		Type:      models.MsgSFTPDownload,
		SessionID: msg.SessionID,
		Data:      map[string]string{"status": "completed"},
	})
}

func (h *SFTPHandler) handleDelete(msg *models.IPCMessage, writer ResponseWriter) error {
	jsonData, err := json.Marshal(msg.Data)
	if err != nil {
		return fmt.Errorf("invalid data: %w", err)
	}

	var req models.DeleteRequest
	if err := json.Unmarshal(jsonData, &req); err != nil {
		return fmt.Errorf("failed to parse delete request: %w", err)
	}

	if err := h.manager.DeleteFile(msg.SessionID, req.Path); err != nil {
		return err
	}

	return writer.WriteMessage(&models.IPCMessage{
		Type:      models.MsgSFTPDelete,
		SessionID: msg.SessionID,
		Data:      map[string]string{"status": "deleted", "path": req.Path},
	})
}

func (h *SFTPHandler) handleMkdir(msg *models.IPCMessage, writer ResponseWriter) error {
	jsonData, err := json.Marshal(msg.Data)
	if err != nil {
		return fmt.Errorf("invalid data: %w", err)
	}

	var req models.MkdirRequest
	if err := json.Unmarshal(jsonData, &req); err != nil {
		return fmt.Errorf("failed to parse mkdir request: %w", err)
	}

	if err := h.manager.CreateDirectory(msg.SessionID, req.Path); err != nil {
		return err
	}

	return writer.WriteMessage(&models.IPCMessage{
		Type:      models.MsgSFTPMkdir,
		SessionID: msg.SessionID,
		Data:      map[string]string{"status": "created", "path": req.Path},
	})
}

func (h *SFTPHandler) handleRename(msg *models.IPCMessage, writer ResponseWriter) error {
	jsonData, err := json.Marshal(msg.Data)
	if err != nil {
		return fmt.Errorf("invalid data: %w", err)
	}

	var req models.RenameRequest
	if err := json.Unmarshal(jsonData, &req); err != nil {
		return fmt.Errorf("failed to parse rename request: %w", err)
	}

	if err := h.manager.RenameFile(msg.SessionID, req.OldPath, req.NewPath); err != nil {
		return err
	}

	return writer.WriteMessage(&models.IPCMessage{
		Type:      models.MsgSFTPRename,
		SessionID: msg.SessionID,
		Data:      map[string]string{"status": "renamed", "old_path": req.OldPath, "new_path": req.NewPath},
	})
}


func (h *SFTPHandler) handleCancel(msg *models.IPCMessage, writer ResponseWriter) error {
	jsonData, err := json.Marshal(msg.Data)
	if err != nil {
		return fmt.Errorf("invalid data: %w", err)
	}

	var req models.CancelRequest
	if err := json.Unmarshal(jsonData, &req); err != nil {
		return fmt.Errorf("failed to parse cancel request: %w", err)
	}

	cancelled := h.manager.CancelTransfer(req.TransferID)

	return writer.WriteMessage(&models.IPCMessage{
		Type:      models.MsgSFTPCancel,
		SessionID: msg.SessionID,
		Data:      map[string]interface{}{"transfer_id": req.TransferID, "cancelled": cancelled},
	})
}


func (h *SFTPHandler) handleReadFile(msg *models.IPCMessage, writer ResponseWriter) error {
	jsonData, err := json.Marshal(msg.Data)
	if err != nil {
		return fmt.Errorf("invalid data: %w", err)
	}

	var req models.ReadFileRequest
	if err := json.Unmarshal(jsonData, &req); err != nil {
		return fmt.Errorf("failed to parse read file request: %w", err)
	}

	content, err := h.manager.ReadFile(msg.SessionID, req.Path, req.Binary)
	if err != nil {
		return err
	}

	return writer.WriteMessage(&models.IPCMessage{
		Type:      models.MsgSFTPReadFile,
		SessionID: msg.SessionID,
		Data:      models.ReadFileResponse{Content: content, Path: req.Path},
	})
}

func (h *SFTPHandler) handleWriteFile(msg *models.IPCMessage, writer ResponseWriter) error {
	jsonData, err := json.Marshal(msg.Data)
	if err != nil {
		return fmt.Errorf("invalid data: %w", err)
	}

	var req models.WriteFileRequest
	if err := json.Unmarshal(jsonData, &req); err != nil {
		return fmt.Errorf("failed to parse write file request: %w", err)
	}

	if err := h.manager.WriteFile(msg.SessionID, req.Path, req.Content); err != nil {
		return err
	}

	return writer.WriteMessage(&models.IPCMessage{
		Type:      models.MsgSFTPWriteFile,
		SessionID: msg.SessionID,
		Data:      map[string]string{"status": "saved", "path": req.Path},
	})
}

func (h *SFTPHandler) handleChmod(msg *models.IPCMessage, writer ResponseWriter) error {
	jsonData, err := json.Marshal(msg.Data)
	if err != nil {
		return fmt.Errorf("invalid data: %w", err)
	}

	var req models.ChmodRequest
	if err := json.Unmarshal(jsonData, &req); err != nil {
		return fmt.Errorf("failed to parse chmod request: %w", err)
	}

	if err := h.manager.Chmod(msg.SessionID, req.Path, req.Mode); err != nil {
		return err
	}

	return writer.WriteMessage(&models.IPCMessage{
		Type:      models.MsgSFTPChmod,
		SessionID: msg.SessionID,
		Data:      map[string]interface{}{"status": "changed", "path": req.Path, "mode": req.Mode},
	})
}

